
Taken from zebra mailing list:

> List:     zebra
> Subject:  [zebra 11365] Re: bgpd segmentation fault
> From:     Sébastien BLON
> Date:     2001-11-26 19:49:37
> 
> The select manpage (2) on Linux says :
> 
>  On  Linux,  timeout  is  modified to reflect the amount of
>        time not slept; most other implementations do not do this.
>        This  causes  problems  both  when  Linux code which reads
>        timeout is ported to other  operating  systems,  and  when
>        code  is  ported to Linux that reuses a struct timeval for
>        multiple selects in  a  loop  without  reinitializing  it.
>        Consider timeout to be undefined after select returns.
> 
> So, I have patched the bgp_next_timer function in bgpd.c as follow :

diff -durN zebra-0.92a.orig/bgpd/bgpd.c zebra-0.92a/bgpd/bgpd.c
--- zebra-0.92a.orig/bgpd/bgpd.c	Sun Aug 19 11:02:38 2001
+++ zebra-0.92a/bgpd/bgpd.c	Fri Apr 12 13:50:09 2002
@@ -7457,9 +7457,14 @@
 int
 bgp_next_timer (struct thread *thread)
 {
+  int time_to_sleep;
   struct timeval timer_now;
   gettimeofday (&timer_now, NULL);
-  return thread->u.sands.tv_sec - timer_now.tv_sec;
+  time_to_sleep=thread->u.sands.tv_sec - timer_now.tv_sec;
+  if(time_to_sleep > 0)
+       return time_to_sleep;
+  else
+       return 1;
 }
 
 void
